connector/sumconnector/sum.go

```diff
@@ -0,0 +1,159 @@
+	return &summer[K]{
+		metricDefs: metricDefs,
+		sums:       make(map[string]map[[16]byte]*attrSummer, len(metricDefs)),
+		timestamp:  time.Now(),
+	}
+}
+
+type summer[K any] struct {
+	metricDefs map[string]metricDef[K]
+	sums       map[string]map[[16]byte]*attrSummer
+	timestamp  time.Time
+}
+
+type attrSummer struct {
+	attrs pcommon.Map
+	sum   float64
+}
+
+func (c *summer[K]) update(ctx context.Context, attrs pcommon.Map, tCtx K) error {
+	var multiError error
+	for name, md := range c.metricDefs {
+		sourceAttribute := md.sourceAttr
+		sumAttrs := pcommon.NewMap()
+		var sumVal float64
+
+		// Get source attribute value
+		if sourceAttrVal, ok := attrs.Get(sourceAttribute); ok {
+			switch {
+			case sourceAttrVal.Str() != "":
+				sumVal, _ = strconv.ParseFloat(sourceAttrVal.Str(), 64)
+			case sourceAttrVal.Double() != 0:
+				sumVal = sourceAttrVal.Double()
+			case sourceAttrVal.Int() != 0:
+				sumVal = float64(sourceAttrVal.Int())
+			}
+		}
+
+		// Get attribute values to include otherwise use default value
+		for _, attr := range md.attrs {
+			if attrVal, ok := attrs.Get(attr.Key); ok {
+				switch {
+				case attrVal.Str() != "":
+					sumAttrs.PutStr(attr.Key, attrVal.Str())
+				case attrVal.Double() != 0:
+					sumAttrs.PutStr(attr.Key, fmt.Sprintf("%v", attrVal.Double()))
+				case attrVal.Int() != 0:
+					sumAttrs.PutStr(attr.Key, fmt.Sprintf("%v", attrVal.Int()))
+				}
+			} else if attr.DefaultValue != nil {
+				switch v := attr.DefaultValue.(type) {
+				case string:
+					if v != "" {
+						sumAttrs.PutStr(attr.Key, v)
+					}
+				case int:
+					if v != 0 {
+						sumAttrs.PutInt(attr.Key, int64(v))
+					}
+				case float64:
+					if v != 0 {
+						sumAttrs.PutDouble(attr.Key, float64(v))
+					}
+				}
+			}
+		}
+
+		// Missing necessary attributes
+		if sumAttrs.Len() != len(md.attrs) {
+			continue
+		}
+
+		// Perform condition matching or not
+		if md.condition == nil {
+			multiError = errors.Join(multiError, c.increment(name, sumVal, sumAttrs))
+			continue
+		}
+
+		if match, err := md.condition.Eval(ctx, tCtx); err != nil {
+			multiError = errors.Join(multiError, err)
+		} else if match {
+			multiError = errors.Join(multiError, c.increment(name, sumVal, sumAttrs))
+		}
+	}
+	return multiError
+}
+
+func (c *summer[K]) increment(metricName string, sumVal float64, attrs pcommon.Map) error {
+	if _, ok := c.sums[metricName]; !ok {
+		c.sums[metricName] = make(map[[16]byte]*attrSummer)
+	}
+
+	key := noAttributes
+	if attrs.Len() > 0 {
+		key = pdatautil.MapHash(attrs)
+	}
+
+	if _, ok := c.sums[metricName][key]; !ok {
+		c.sums[metricName][key] = &attrSummer{attrs: attrs}
+	}
+
+	for strings := range c.sums[metricName][key].attrs.AsRaw() {
+		if _, ok := c.sums[metricName][key].attrs.Get(strings); ok {
+			c.sums[metricName][key].sum += sumVal
+		}
+	}
+
+	if attrs.Len() == 0 {
+		c.sums[metricName][key].sum += sumVal
+	}
+
+	return nil
+}
+
+func reduceFloatPrecision(val float64) float64 {
```

Comment:
It'd be good to add a comment here about why this is necessary.

Reply:
Added one for clarification this addresses float precision below 6 decimals.

Or would you like me to explicitly call out the testing and add a TODO: add float threshold comparison option to `CompareMetrics`? Just let me know and I'm happy to do that.

Reply:
Is there a reason this is necessary in the functionality of the component, instead of just being used for tests? My original understanding was it would just be used for comparing test values, I didn't think we'd need this when setting metric values.

Reply:
Oh agree!
This is primarily here so you can check out the logic and the tests can pass as is.

The issue as I see it is currently `CompareMetrics` func can't ignore float precision and to do that comparison with float precision threshold I'd need to basically duplicate the entire CompareMetrics functions (the nested parsing especially) and add the float precision thresholding function. That or put this all on hold to PR that into the main repo's testing options.

Unless there is some more creative way to do this comparing of two metric values that is eluding me (please say there is!)

Reply:
Added an issue and will happily add an option to ignore tiny deviations of float values to `pmetrictest` once the issue has been triaged/assigned/etc. 👍

Reply:
ok we're using the new option to ignore float precision after 3 decimals.